### PR TITLE
[DROOLS-7584] process completely all facts inserted before halt

### PR DESCRIPTION
**Pull request by mariofusco**
**JIRA**: https://issues.redhat.com/browse/DROOLS-7584

### CODE REVIEW

**Comment by kie-ci3** [comment]
**PR job** `#1` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-drools/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-drools -u https://github.com/apache/incubator-kie-drools/pull/5576 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-5576/1/display/redirect

**Test results:**
- PASSED: 21738
- FAILED: 2

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-5576/1/testReport/org.drools.model.codegen.execmodel/FromTest/testModifyWithFromSudoku[STANDARD_FROM_DRL]/">org.drools.model.codegen.execmodel.FromTest.testModifyWithFromSudoku[STANDARD_FROM_DRL]</a></summary>
<br/>expected: null<br/> but was: 3
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-5576/1/testReport/org.drools.model.codegen.execmodel/FromTest/testModifyWithFromSudoku[PATTERN_DSL]/">org.drools.model.codegen.execmodel.FromTest.testModifyWithFromSudoku[PATTERN_DSL]</a></summary>
<br/>expected: null<br/> but was: 3
</details>

**Comment by tkobayas** [comment]
Hi @mariofusco 

I sent https://github.com/mariofusco/drools/pull/22 to fix the test failure.

Actually, Sudoku example uses `drools.halt()` to stop the evaluation. The test case can be fixed with the above PR, but the Sudoku example itself is not yet fixed, because it's a little more complicated, so it would require another kie-issue to fix the Sudoku example (With your PR, Sudoku immediately displays a full answer. "Step" button doesn't work).

That said, this looks like a little breaking change. Some users may be affected, but I guess it's a very minor use case, so maybe okay to go.

File: drools-core/src/main/java/org/drools/core/impl/ActivationsManagerImpl.java [hunk 1]
```diff
@@ -59,6 +52,13 @@
 import org.kie.api.event.rule.MatchCancelledCause;
 import org.kie.api.runtime.rule.AgendaFilter;
 
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
```
Comment (tkobayas) [inline drools-core/src/main/java/org/drools/core/impl/ActivationsManagerImpl.java]: Hi @mariofusco , you may be already aware, the order of import is against checkstyle (= currently Paolo is working on). Don't you use https://github.com/kiegroup/droolsjbpm-build-bootstrap/blob/main/ide-configuration/intellij-configuration/code-style/intellij-code-style_droolsjbpm-java-conventions.xml ?

**Review by lucamolteni — APPROVED** [review]

File: drools-core/src/main/java/org/drools/core/impl/ActivationsManagerImpl.java (hunk 1 above)
Comment (mariofusco) [reply drools-core/src/main/java/org/drools/core/impl/ActivationsManagerImpl.java]: @tkobayas thanks, fixed.

**Comment by kie-ci3** [comment]
**PR job** `#2` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-drools/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-drools -u https://github.com/apache/incubator-kie-drools/pull/5576 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-5576/2/display/redirect

**Test results:**
- PASSED: 3772
- FAILED: 2

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-5576/2/testReport/org.drools.model.codegen.execmodel/FromTest/testModifyWithFromSudoku[STANDARD_FROM_DRL]/">org.drools.model.codegen.execmodel.FromTest.testModifyWithFromSudoku[STANDARD_FROM_DRL]</a></summary>
<br/>expected: null<br/> but was: 3
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-5576/2/testReport/org.drools.model.codegen.execmodel/FromTest/testModifyWithFromSudoku[PATTERN_DSL]/">org.drools.model.codegen.execmodel.FromTest.testModifyWithFromSudoku[PATTERN_DSL]</a></summary>
<br/>expected: null<br/> but was: 3
</details>

**Review by tkobayas — APPROVED** [review]

**Pull request merged**